Classes/Hooks/PageLayoutController.php

```diff
@@ -32,116 +32,115 @@
+            $jsLines[] = 'top.pasteReferenceAllowed = ' . (int)$this->helper->getBackendUser()->checkAuthMode('tt_content', 'CType', 'shortcut') . ';';
+            $jsLines[] = 'top.browserUrl = ' . json_encode((string)$uriBuilder->buildUriFromRoute('wizard_element_browser')) . ';';
         } catch (RouteNotFoundException $e) {
         }
 
-        if (!empty($elFromTable)) {
-            $pasteItem = (int)substr((string)key($elFromTable), 11);
-            $pasteRecord = BackendUtility::getRecordWSOL('tt_content', $pasteItem) ?? [];
-            $pasteTitle = BackendUtility::getRecordTitle('tt_content', $pasteRecord);
-
-            if (!(bool)($this->extensionConfiguration['disableCopyFromPageButton'] ?? false)
-                && !(bool)($this->helper->getBackendUser()->uc['disableCopyFromPageButton'] ?? false)
-            ) {
-                /* @todo: the file @haerer/paste-reference/paste-reference.js doesn't exist.
-                 * @todo: It should add a button next to "+ Content"
-                 * $this->pageRenderer->getJavaScriptRenderer()->addJavaScriptModuleInstruction(
-                 * JavaScriptModuleInstruction::create('@haerer/paste-reference/paste-reference.js')
-                 * ->instance([
-                 * 'itemOnClipboardUid' => $pasteItem,
-                 * 'itemOnClipboardTitle' => $pasteTitle,
-                 * 'copyMode' => $clipboard->clipData['normal']['mode'] ?? '',
-                 * ])
-                 * );*/
-
-                $pAddExtOnReadyCode .= '
-                    top.copyFromAnotherPageLinkTemplate = ' . json_encode('<button type="button" class="t3js-paste-new btn btn-default" title="' . ($this->helper->getLanguageService()?->sL('LLL:EXT:paste_reference/Resources/Private/Language/locallang_db.xml:tx_paste_reference_js.copyfrompage') ?? 'unknown') . '">' . $this->iconFactory->getIcon(
-                    'actions-insert-reference',
-                    Icon::SIZE_SMALL
-                )->render() . '</button>') . ';';
-            }
+        if (!empty($this->elFromTable)) {
+            $this->addJavaScriptModuleInstruction();
+            $jsLines[] = 'top.copyMode = "' . $this->copyMode . '";';
         }
 
-        $this->pageRenderer->addJsInlineCode(
-            'pasterefExtOnReady',
-            $pAddExtOnReadyCode,
-            true,
-            false,
-            true
-        );
+        if (
+            !(bool)($this->extensionConfiguration['disableCopyFromPageButton'] ?? false)
+            && !(bool)($this->helper->getBackendUser()->uc['disableCopyFromPageButton'] ?? false)
+        ) {
+            $jsLines[] = 'top.copyFromAnotherPageLinkTemplate = ' . json_encode($this->getButtonTemplate()) . ';';
+        }
+
+        $javaScript = implode("\n", $jsLines);
+        $this->pageRenderer->addJsInlineCode('pasteReference', $javaScript, true, false, true);
```

Comment:
inline code in general is a bad thing, specially with (upcoming) CSP. 

Would have loved to investigate this how to avoid inline code at all and verify this against active CSP rules firste - which was prevented by a unapproved merge.

Should be done before working on v13 or making a release.

Reply:
using `setTimeout` is prevented by CSP.
having a look ...

Reply:
I think it's feature request to  change this and should be posted as issue.

Reply:
@sbuerk  I just get doubts the the inline code is still required as I included to fetch the clipboard by Ajax.
So required changes beside code-removal might be marginal.
I'll check it out.

Reply:
@DavidBruchmann Did you try $this->pageRenderer->addJsFile() instead?